noir-projects/aztec-nr/aztec/src/macros/dispatch/mod.nr

```diff
@@ -0,0 +1,141 @@
+use super::utils::compute_fn_selector;
+
+/// Returns an `fn public_dispatch(...)` function for the given module that's assumed to be an Aztec contract.
+pub comptime fn generate_public_dispatch(m: Module) -> Quoted {
+    let functions = m.functions();
+    let functions = functions.filter(|function: FunctionDefinition| function.has_named_attribute("public"));
+
+    let unit = get_type::<()>();
+
+    let ifs = functions.map(
+        |function: FunctionDefinition| {
+        let name = function.name();
+        let parameters = function.parameters();
+        let return_type = function.return_type();
+
+        let selector: Field = compute_fn_selector(function);
+
+        let mut parameters_size = 0;
+        for param in parameters {
+            parameters_size += size_in_fields(param.1);
+        }
+
+        let initial_read = if parameters.len() == 0 {
+            quote {}
+        } else {
+            // The initial calldata_copy offset is 1 to skip the Field selector
+            quote {
+                let input_calldata: [Field; $parameters_size] = dep::aztec::context::public_context::calldata_copy(1, $parameters_size);
+                let mut reader = dep::aztec::protocol_types::utils::reader::Reader::new(input_calldata);
+            }
```

Comment:
It'd be nice to have a comment explaining the expected calldata layout, maybe with an example of a concrete fn with a given signature

Reply:
Done.